### ISSUE TITLE
Proposal to eliminate polling in chord_unlock with RabbitMQ

### ISSUE THREAD

**Issue by averwind** [celery/celery#2748]
Right now the chord unlock task checks each task in group/header and than retries itself. What we thought about is that every member of the group will know the total number of tasks (N) in the group and a separate queue will be created holding indices from 1 to N in order. As each task finished it will pull an item from this queue if it is N it will call the callback, if it is not it will just remove the message it consumed from the queue. 

Wanted to see what the community thinks about this approach before making a pull request.

**Comment by thedrow** [comment]
What could be the implications on existing users?
What do you mean that the queue will hold indices in order? To the best of my knowledge ordering in RabbitMQ is guaranteed on a best effort basis.

**Comment by averwind** [comment]
That's a good point, as an alternative what I had in mind was to put N-1 tokens in the queue so the last one to try to get a token and see that queue is empty will make the call to callback.

**Comment by thedrow** [comment]
I think that could work. But we need to keep the old behavior the default for now.

**Comment by ask** [comment]
That's a neat idea, I think it could work if each task removed the message immediately (no_ack) and
if receiving the Nth token, trying to consume another token to verify that it's the last one.

The only way to find out if it works is to actually implement it and test it thoroughly :)

**Comment by ask** [comment]
Maybe instead of consuming another token with no_ack=1, it could do a ack the next time,

    token1 = channel.basic_get(chord_queue, no_ack=True)
    if token1.payload == callback.chord_size:
       token2 = channel.basic_get(chord_queue, no_ack=False)
       if token2:
           
          call_callback_task()

**Comment by ask** [comment]
Well, you cannot actually use any kind of verification, because then another task could have consumed it while you're trying to verify.  So it would have to trust the ordering.  The big downside to this approach, is that if you want a chord of 10k tasks you also have to fill the queue with 10k tokens, but that will not work with the current implementation anyway so I guess it's not much of an issue :)

**Comment by averwind** [comment]
Right, the ordering was a problem as solution to that I was thinking instead of relying on ordering we put N-1 tokens into the queue, if you can acquire a token at the end of the process that's too bad as it means group is not done yet but if the queue is empty, great you were the last task and you get to make the callback call.

**Comment by sireeshk** [comment]
There could be a race when acks_late = TRUE for the tasks in the group and if the worker crashes after consuming the token but before sending the ack. This task will be re-tried and could end up calling the callback before all tasks in the group are complete.

**Comment by thedrow** [comment]
In that case we can implement this behavior only if ack late is set to false.
